chain/client/src/stateless_validation/partial_witness/partial_witness_actor.rs

```diff
@@ -566,6 +602,37 @@ impl PartialWitnessActor {
         validators.sort();
         Ok(validators)
     }
+
+    /// Returns for each code-hash the contract code retrieved from the storage.
+    fn retrieve_contract_code<'a, I>(
+        &self,
+        key: &ChunkProductionKey,
+        code_hashes: I,
+    ) -> Result<Vec<CodeBytes>, Error>
+    where
+        I: Iterator<Item = &'a CodeHash>,
+    {
+        let storage = TrieDBStorage::new(
+            TrieStoreAdapter::new(self.runtime.store().clone()),
+            self.epoch_manager.shard_id_to_uid(key.shard_id, &key.epoch_id)?,
+        );
+        let mut contracts = Vec::new();
+        for contract_hash in code_hashes {
+            match storage.retrieve_raw_bytes(&contract_hash.0) {
+                Ok(bytes) => contracts.push(CodeBytes(bytes)),
+                Err(StorageError::MissingTrieValue(_, _)) => {
+                    tracing::warn!(
```

Comment:
Is this failure weaker than an error?

Reply:
ideally should be error (should never happen), but it might be the case that the requester adds invalid code hashes.

Reply:
@pugachAG fyi.